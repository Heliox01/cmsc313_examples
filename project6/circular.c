#include "circular.h"

/*
 * This is where you should write all your code (except any additional
 * tests you choose to include).
 */


/*
 * First define what the real _cbuf struct should look like. Remember,
 * this is opaque to the users of your library so you can do whatever
 * you like in here. The _cbuf struct is typedefd in the header to the
 * cbuf type so you're free to use either type mostly
 * interchangably. The one exception is that sizeof calls must use the
 * _cbuf type.
 */

struct _cbuf {
  // whatever you want can go in here
};

/*
 * Now define your implementations of all of the functions given in
 * the header. The test programs will not link successfully until each
 * function they use is defined here.
 */
