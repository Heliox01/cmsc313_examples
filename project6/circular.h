#ifndef CIRCULAR_H
#define CIRCULAR_H

/* provides size_t */
#include <stddef.h>

/*
 * This is an opaque data structure. You may define the _cbuf struct
 * inside your C file and include whatever fields you like. The user
 * of your library should never know or care about the internals of
 * your cbuf struct.
 */
struct _cbuf;
typedef struct _cbuf cbuf;

/*
 * This is the type that users will store inside your data
 * structure. Its definition is exposed to users because they will
 * need to be able to access its fields.
 *
 * Your implementation should not need to modify the provided
 * definition of this datastructure.
 */
typedef struct {
  double rate;
  unsigned int time;
} quote;


/*
 * These are the "public" functions used for interacting with your
 * datastructure. You are free to define more functions to be used as
 * internal implementation details but you should define them only in
 * your C file so that the users of your datastructure will not need
 * to know or care about them. If you wish, you can create a seperate
 * "circular_internal.h" header that has definitions of these internal
 * functions so that your tests can access them.
 */


/*
 * Initialize a circular buffer and return a pointer that can be used
 * to specify that circular buffer in future function calls.
 */
cbuf* cbuf_alloc();

/*
 * Deallocate all dynamically allocated memory associated with the
 * circular buffer referenced by cb_ptr. The programmer using your ADT
 * promises to never use the value in cb_ptr again.
 */
void cbuf_free(cbuf *cb_ptr);

/*
 * Add a new currency exchange quote to the specified circular
 * buffer. The time is in seconds since midnight. This function should
 * remove any quotes in the circular buffer that are more than 5
 * minutes older than this quote. You should assume that the
 * timestamps given to cbuf_update() are in non-decreasing
 * order. During a cbuf_update(), the memory for the circular buffer
 * may have to grow or shrink according to the rules given in the
 * project description.
 */
void cbuf_update(cbuf *cb_ptr, unsigned int time, double rate);

/*
 * Return the average exchange rate of the quotes currently stored in
 * the buffer.
 */
double cbuf_average(cbuf *cb_ptr);

/*
 * Return a pointer to the earliest quote currently stored in the
 * buffer. (Earliest = smallest timestamp) The programmer using your
 * ADT promises to just "look at" the quote and not alter it in any
 * way.
 */
const quote *cbuf_start(cbuf *cb_ptr);

/*
 * Return a pointer to the latest quote currently stored in the
 * buffer. (Latest = largest timestamp) The programmer using your ADT
 * promises to just "look at" the quote and not alter it in any way.
 */
const quote *cbuf_end(cbuf *cb_ptr);

/*
 * Return the number of elements currently stored in the circular
 * buffer.
 */
size_t cbuf_size(cbuf *cb_ptr);

/*
 * Return the current capacity of the circular buffer (will be greater
 * than or equal to the number of elements in the buffer). When
 * cbuf_size(x) == cbuf_capacity(x) then your circular buffer may need
 * to grow when the next element is inserted.
 */
size_t cbuf_capacity(cbuf *cb_ptr);

#endif
