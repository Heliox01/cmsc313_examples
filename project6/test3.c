/*
 * Test out circular buffer for foreign exchange quotes.
*/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "circular.h"


// Print out some useful data in the given circular buffer.
//
void print_data(cbuf *cb_ptr) {
  double avg;
  avg = cbuf_average(cb_ptr);
  printf("Average rate = %f\n", avg);

  const quote *q_start, *q_end;
  q_start = cbuf_start(cb_ptr);
  q_end = cbuf_end(cb_ptr);

  printf("Start: time = %d, rate = %f\n", q_start->time, q_start->rate);
  printf("End:   time = %d, rate = %f\n", q_end->time, q_end->rate);

  return;
}


int main(int argc, char** argv) {
  cbuf *eur_usd, *eur_jpy, *usd_jpy;
  FILE *ifile;
  char* fname;
  char pair[10];
  unsigned int hour, min, isec, time;
  double sec, rate;
  int r;

  eur_usd = cbuf_alloc();
  eur_jpy = cbuf_alloc();
  usd_jpy = cbuf_alloc();

  if(argc != 2) {
    fprintf(stderr, "usage: %s filename\n", argv[0]);
    exit(1);
  }

  fname = argv[1];

  // Try to open file
  ifile = fopen (fname, "r");
  if (ifile == NULL) {
    fprintf(stderr, "Could not open file: %s\n", fname);
    exit(1);
  }


  // keep reading file until no more ticks
  //
  while(1) {

    r = fscanf(ifile, "%7s %d:%d:%lf %lf", pair, &hour, &min, &sec, &rate);
    isec = sec;

    // fprintf(stderr, "matched %d items\n", r);
    // fprintf(stderr, "%s, %02d:%02d:%02d %f\n", pair, hour, min, isec, rate);

    time = isec + 60 * min + 3600 * hour;


    if (r <= 0) break;   // EOF?

    if (strcmp(pair, "EUR-USD") == 0) {

      cbuf_update(eur_usd, time, rate);

    } else if (strcmp(pair, "EUR-JPY") == 0) {

      cbuf_update(eur_jpy, time, rate);

    } else if (strcmp(pair, "USD-JPY") == 0) {

      cbuf_update(usd_jpy, time, rate);

    } else {
      fprintf(stderr, "Not a recognized currency pair!\n");
    }

  }

  printf("\n\n*** Data for EUR-USD ***\n");
  print_data(eur_usd);

  printf("\n\n*** Data for EUR-JPY ***\n");
  print_data(eur_jpy);

  printf("\n\n*** Data for USD-JPY ***\n");
  print_data(usd_jpy);


  // Be nice, clean up
  cbuf_free(eur_usd);
  cbuf_free(eur_jpy);
  cbuf_free(usd_jpy);
  fclose(ifile) ;

  return 0;
}
