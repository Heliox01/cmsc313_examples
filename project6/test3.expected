

*** Data for EUR-USD ***
Average rate = 0.897167
Start: time = 22025, rate = 1.304240
End:   time = 22313, rate = 1.304140


*** Data for EUR-JPY ***
Average rate = 123.603972
Start: time = 22020, rate = 123.609000
End:   time = 22319, rate = 123.534000


*** Data for USD-JPY ***
Average rate = 94.754715
Start: time = 22020, rate = 94.774000
End:   time = 22319, rate = 94.724000
